src/idmap/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+#![deny(warnings)]
+#![warn(unused_extern_crates)]
+#![deny(clippy::todo)]
+#![deny(clippy::unimplemented)]
+#![deny(clippy::unwrap_used)]
+#![deny(clippy::expect_used)]
+#![deny(clippy::panic)]
+#![deny(clippy::unreachable)]
+#![deny(clippy::await_holding_lock)]
+#![deny(clippy::needless_pass_by_value)]
+#![deny(clippy::trivially_copy_pass_by_ref)]
+use std::ffi::CString;
+use std::ptr;
+use std::sync::RwLock;
+use uuid::Uuid;
+
+#[macro_use]
+extern crate tracing;
+
+mod ffi {
+    #![allow(non_upper_case_globals)]
+    #![allow(non_camel_case_types)]
+    #![allow(non_snake_case)]
+    #![allow(dead_code)]
+    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+}
+
+#[allow(non_camel_case_types)]
+pub enum IdmapError {
+    IDMAP_SUCCESS,
+    IDMAP_NOT_IMPLEMENTED,
+    IDMAP_ERROR,
+    IDMAP_OUT_OF_MEMORY,
+    IDMAP_NO_DOMAIN,
+    IDMAP_CONTEXT_INVALID,
+    IDMAP_SID_INVALID,
+    IDMAP_SID_UNKNOWN,
+    IDMAP_NO_RANGE,
+    IDMAP_BUILTIN_SID,
+    IDMAP_OUT_OF_SLICES,
+    IDMAP_COLLISION,
+    IDMAP_EXTERNAL,
+    IDMAP_NAME_UNKNOWN,
+    IDMAP_ERR_LAST,
+}
+
+fn map_err(e: ffi::idmap_error_code) -> IdmapError {
+    match e {
+        ffi::idmap_error_code_IDMAP_SUCCESS => IdmapError::IDMAP_SUCCESS,
+        ffi::idmap_error_code_IDMAP_NOT_IMPLEMENTED => IdmapError::IDMAP_NOT_IMPLEMENTED,
+        ffi::idmap_error_code_IDMAP_ERROR => IdmapError::IDMAP_ERROR,
+        ffi::idmap_error_code_IDMAP_OUT_OF_MEMORY => IdmapError::IDMAP_OUT_OF_MEMORY,
+        ffi::idmap_error_code_IDMAP_NO_DOMAIN => IdmapError::IDMAP_NO_DOMAIN,
+        ffi::idmap_error_code_IDMAP_CONTEXT_INVALID => IdmapError::IDMAP_CONTEXT_INVALID,
+        ffi::idmap_error_code_IDMAP_SID_INVALID => IdmapError::IDMAP_SID_INVALID,
+        ffi::idmap_error_code_IDMAP_SID_UNKNOWN => IdmapError::IDMAP_SID_UNKNOWN,
+        ffi::idmap_error_code_IDMAP_NO_RANGE => IdmapError::IDMAP_NO_RANGE,
+        ffi::idmap_error_code_IDMAP_BUILTIN_SID => IdmapError::IDMAP_BUILTIN_SID,
+        ffi::idmap_error_code_IDMAP_OUT_OF_SLICES => IdmapError::IDMAP_OUT_OF_SLICES,
+        ffi::idmap_error_code_IDMAP_COLLISION => IdmapError::IDMAP_COLLISION,
+        ffi::idmap_error_code_IDMAP_EXTERNAL => IdmapError::IDMAP_EXTERNAL,
+        ffi::idmap_error_code_IDMAP_NAME_UNKNOWN => IdmapError::IDMAP_NAME_UNKNOWN,
+        ffi::idmap_error_code_IDMAP_ERR_LAST => IdmapError::IDMAP_ERR_LAST,
+        _ => {
+            error!("Unknown error code '{}'", e);
+            IdmapError::IDMAP_ERROR
+        }
+    }
+}
+
+fn object_id_to_sid(object_id: &Uuid) -> Result<String, IdmapError> {
```

Comment:
> Do we know how windows turns an azure ad uuid into a SID? That could be what we need to re-create actually.

This is how. I created this function based on Windows behavior when joined to azure ad.

Reply:
From that could we re-implement the idmap SID -> UID process then in here?

Reply:
I believe so, and that's probably what I'll end up doing. I've taken a break on this one and I'm working on a more critical bug at the moment.